helix-core/src/indent.rs

```diff
@@ -459,40 +492,119 @@ fn query_indents(
 ///     },
 /// );
 /// ```
+#[allow(clippy::too_many_arguments)]
 pub fn treesitter_indent_for_pos(
     query: &Query,
     syntax: &Syntax,
     indent_style: &IndentStyle,
+    tab_width: usize,
     text: RopeSlice,
     line: usize,
     pos: usize,
     new_line: bool,
 ) -> Option<String> {
     let byte_pos = text.char_to_byte(pos);
+    // The innermost tree-sitter node which is considered for the indent
+    // computation. It may change if some predeceding node is extended
     let mut node = syntax
         .tree()
         .root_node()
         .descendant_for_byte_range(byte_pos, byte_pos)?;
-    let mut first_in_line = get_first_in_line(node, byte_pos, new_line);
-    let new_line_break = if new_line {
-        Some((line, byte_pos))
-    } else {
-        None
-    };
-    let query_result = crate::syntax::PARSER.with(|ts_parser| {
+    let (query_result, prev_child) = crate::syntax::PARSER.with(|ts_parser| {
         let mut ts_parser = ts_parser.borrow_mut();
         let mut cursor = ts_parser.cursors.pop().unwrap_or_else(QueryCursor::new);
+        // The query range should intersect with all nodes directly preceding
+        // the cursor in case one of them is extended.
+        // prev_child is the deepest such node.
+        let (query_range, prev_child) = {
+            // TODO Is there some way we can reuse this cursor?
+            let mut tree_cursor = node.walk();
+            let mut prev_child = None;
+            for child in node.children(&mut tree_cursor) {
+                if child.byte_range().end <= byte_pos {
+                    prev_child = Some(child);
+                }
+            }
+            match prev_child {
+                Some(mut prev_child) => {
+                    // Get the deepest directly preceding node
+                    while prev_child.child_count() > 0 {
+                        prev_child = prev_child.child(prev_child.child_count() - 1).unwrap();
+                    }
+                    (
+                        prev_child.byte_range().end - 1..byte_pos + 1,
+                        Some(prev_child),
+                    )
+                }
+                None => (byte_pos..byte_pos + 1, None),
+            }
+        };
         let query_result = query_indents(
             query,
             syntax,
             &mut cursor,
             text,
-            byte_pos..byte_pos + 1,
-            new_line_break,
+            query_range,
+            new_line.then(|| (line, byte_pos)),
         );
         ts_parser.cursors.push(cursor);
-        query_result
+        (query_result, prev_child)
     });
+    let indent_captures = query_result.indent_captures;
+    let extend_captures = query_result.extend_captures;
+
+    // Check for extend captures (starting with the deepest
+    // candidate node and then going up the syntax tree).
+    if let Some(mut prev_child) = prev_child {
```

Comment:
This block is a bit long and seems to be somewhat well-contained. It may clean this function up a bit to split this into its own function. This function is already quite large and hard to follow